src/main/java/graphql/execution/values/legacycoercing/LegacyCoercingInputInterceptor.java

```diff
@@ -0,0 +1,179 @@
+        return migratesValues((input, type) -> {
+        });
+    }
+
+    /**
+     * This will change legacy values as it encounters them to something acceptable to the more strict coercion rules.
+     * The observer callback will be invoked if it detects a legacy value that it will change.
+     *
+     * @param observerCallback a callback allowing you to observe a legacy scalar value before it is migrated
+     *
+     * @return an InputInterceptor that both observes values and migrates them to a more strict value
+     */
+    public static LegacyCoercingInputInterceptor migratesValues(BiConsumer<Object, GraphQLInputType> observerCallback) {
+        return new LegacyCoercingInputInterceptor(((input, graphQLInputType) -> {
+            observerCallback.accept(input, graphQLInputType);
+            if (Scalars.GraphQLBoolean.equals(graphQLInputType)) {
+                return coerceLegacyBooleanValue(input);
+            }
+            if (Scalars.GraphQLFloat.equals(graphQLInputType)) {
+                return coerceLegacyFloatValue(input);
+            }
+            if (Scalars.GraphQLInt.equals(graphQLInputType)) {
+                return coerceLegacyIntValue(input);
+            }
+            if (Scalars.GraphQLString.equals(graphQLInputType)) {
+                return coerceLegacyStringValue(input);
+            }
+            return input;
+        }));
+    }
+
+    private final BiFunction<Object, GraphQLInputType, Object> behavior;
+
+    private LegacyCoercingInputInterceptor(BiFunction<Object, GraphQLInputType, Object> behavior) {
+        this.behavior = assertNotNull(behavior);
+    }
+
+    @Override
+    public Object intercept(@Nullable Object input, @NotNull GraphQLInputType graphQLType, @NotNull GraphQLContext graphqlContext, @NotNull Locale locale) {
+        if (isLegacyValue(input, graphQLType)) {
+            // we ONLY apply the new behavior IF it's an old acceptable legacy value.
+            // so for compliant values - we change nothing and invoke no behaviour
+            // and for values that would not coerce anyway, we also invoke no behavior
+            return behavior.apply(input, graphQLType);
+        }
+        return input;
+    }
+
+    @SuppressWarnings("RedundantIfStatement")
+    static boolean isLegacyValue(Object input, GraphQLInputType graphQLType) {
+        if (Scalars.GraphQLBoolean.equals(graphQLType)) {
+            return isLegacyBooleanValue(input);
+        } else if (Scalars.GraphQLFloat.equals(graphQLType)) {
+            return isLegacyFloatValue(input);
+        } else if (Scalars.GraphQLInt.equals(graphQLType)) {
+            return isLegacyIntValue(input);
+        } else if (Scalars.GraphQLString.equals(graphQLType)) {
+            return isLegacyStringValue(input);
+        } else {
+            return false;
+        }
+    }
+
+    static boolean isLegacyBooleanValue(Object input) {
+        return input instanceof String || CoercingUtil.isNumberIsh(input);
```

Comment:
I think this PR should be merged as is. One teeny nit in case someone is reading this PR in the future

On naming, technically this method is "is maybe legacy boolean", because only the strings "true" and "false" are acceptable legacy string values.

Retaining this function is perfectly fine as the usual boolean coercion that happens after this interceptor will catch strings that are not either "true" or "false"